Tests/LogtoClientTests/LogtoAuthSession/LogtoAuthSessionTests.swift

```diff
@@ -0,0 +1,87 @@
+import AuthenticationServices
+import Logto
+@testable import LogtoClient
+import XCTest
+
+private let redirectUri = URL(string: "io.logto.test://callback")!
+
+private struct LogtoWebAuthSessionMock: LogtoWebAuthSession {
+    let url: URL
+    let completionHandler: (URL?, Error?) -> Void
+
+    init(url: URL, callbackURLScheme _: String?, completionHandler: @escaping (URL?, Error?) -> Void) {
+        self.url = url
+        self.completionHandler = completionHandler
+    }
+
+    func start() -> Bool {
+        guard url.path != "/canceled" else {
+            completionHandler(nil, ASWebAuthenticationSessionError(.canceledLogin))
+            return true
+        }
+
+        completionHandler(redirectUri, nil)
```

Comment:
nit: would it be better to use a positive statement if possible?:
```
guard url.path = "/canceled" else {
  completionHandler(redirectUri, nil);
  return truel
}

completionHandler(nil, ASWebAuthenticationSessionError(.canceledLogin))
 return true
```

Reply:
yes, but we got more conditions to go. i'll update to switch clause

Reply:
updated, PTAL